doc/python-type-hinting.md

```diff
@@ -64,24 +60,98 @@ def test(self, var: Union[None, int, str]) -> None:
         return [value + 1 for value in values]
 ```
 
-## Forward declarations
+## Self Type
 
-To reference a type before it has been declared e.g. using a class as a type within the class declaration, add the special `annotations` import:
+To annotate methods that return an instance of their class, use the `Self` type is as it is bound to it's encapsulating class. In the example below, the type checker will correctly infer the type of `Circle().set_scale(0.5)` to be `Circle`:
 
 ```python
-from __future__ import annotations
+from typing import Self
+
+class Shape:
+    def set_scale(self, scale: float) -> Self:
+        self.scale = scale
+        return self
 
-class TestClass:
 
-  def test(self, var: Sequence[TestClass]) -> None:
+class Circle(Shape):
+    def set_radius(self, radius: float) -> Self:
+        self.radius = radius
+        return self
+```
+
+This is recommended as forward declarations are now redundant in 3.10
+https://peps.python.org/pep-0673/
+
+## Type Alias
+
+Use the special annotation `TypeAlias` to declare type aliases more explicitly so type checkers are able to distinguish between type aliases and ordinary assignments:
+
+```python
+MyType: TypeAlias = "ClassName"
+def foo() -> MyType: ...
 ```
 
-This import is not necessary in Python 3.10.
+## Type Ignore
+
+To mark portions of the program that should not be covered by type hinting, use `# type: ignore` on the specific line. When used in a line by itself at the top of a file, it silences all errors in the file:
+
+```python
+# type: ignore
+```
+
+`# type: ignore` should only be used when unavoidable. Ensure that a comment is added to explain why it has been used and have a prefix of `Type ignore:`
+
+```python
+# Type ignore: Explain why type ignore has been used in this context
+x: int
+x: str  # type: ignore
```

Comment:
We should add a more meaningful example. This to me isn't valid. Maybe one where it returns any then you explicitly type it.